### PR TITLE
RATEST-277: Replace assertText() with assertPage() to improve the tests run

### PR DESCRIPTION
**Pull request by kdaud**
Ticket [ID](https://issues.openmrs.org/browse/RATEST-277)

Description -> Replace assertText() with assertPage() to improve the test run.

### CODE REVIEW

File: src/test/java/org/openmrs/contrib/qaframework/automation/AllergiesSteps.java [hunk 1]
```diff
@@ -42,27 +42,28 @@ public void loadAllergiesPage() {
...
 
 	@And("a user clicks No Known Allergy button")
 	public void addNoKnownAllergy() {
 		allergyPage.addNoKnownAllergy();
+		allergyPage.waitForPage();
 	}
 
 	@Then("the system add no known allergies into the allergies table")
 	public void systemAddsNoKnownAllergies() {
-		assertTrue(textExists("No known allergies"));
```
Comment (sherrif10) [inline src/test/java/org/openmrs/contrib/qaframework/automation/AllergiesSteps.java:56]: Havent seen a good reason for this refactor, Do you mind giving me a brief way why this change.

Comment (kdaud) [reply src/test/java/org/openmrs/contrib/qaframework/automation/AllergiesSteps.java:56]: Have been noticing random breakage of these automated tests on ci and opted to improve them as a way of ensuring they are stable and can serve the intended purpose otherwise if these tests are not stable then it will have less value as far as our vision is concerned. We need to have them stabilize and reliable on every build plan run !

@dkayiwa  emphasized this via slack where he said `this instability will make it unreliable to be of value`.

Comment (kdaud) [reply src/test/java/org/openmrs/contrib/qaframework/automation/AllergiesSteps.java:56]: > Havent seen a good reason for this refactor

assertPage() is more reliable than assertText() in reference to automated tests.

Comment (kdaud) [reply src/test/java/org/openmrs/contrib/qaframework/automation/AllergiesSteps.java:56]: Have been noticing random breakage of these automated tests on ci and opted to improve them as a way of ensuring they are stable and can serve the intended purpose otherwise if these tests are not stable then it will have less value as far as our vision is concerned. We need to have them stabilize and reliable on every build plan run !

@dkayiwa  emphasized this via slack where he said `this instability will make it unreliable to be of value`.

**Comment by sherrif10** [comment]
@kdaud  does the invisibility of an element comes after page loads or elements

**Comment by kdaud** [comment]
> does the invisibility of an element comes after page loads or elements

I seem not to get your question well, do you mind phrasing it differently?

**Comment by kdaud** [comment]
@sherrif10 have taken a look at the pardon [above](https://github.com/openmrs/openmrs-contrib-qaframework/pull/256#issuecomment-986909772)?

File: src/test/java/org/openmrs/contrib/qaframework/automation/ConditionsSteps.java [hunk 2]
```diff
@@ -64,7 +64,7 @@ public void clickReturn() {
 
 	@Then("System returns to patient dashboard")
 	public void returnToDashboard() {
-		matchPatientIds(patientDashboardId);
+		assertPage(dashboardPage.waitForPage());
```
Comment (sherrif10) [inline src/test/java/org/openmrs/contrib/qaframework/automation/ConditionsSteps.java:67]: So ideally, let's use this example, suppose the system doesn't return to the dashboard page, and your assertion is waiting for dashboard page instead of a functional  method that directly checks on the known element or id , how would you include that assertion

Comment (kdaud) [reply src/test/java/org/openmrs/contrib/qaframework/automation/ConditionsSteps.java:67]: When the system doesn't return to the dashboard page the test must fail the run because the run is expected to be on the dashboard page for the next operation to be successful. As well the previous assertion is good enough to assert the return of the system on the dashboard  since it asserts the patientDashboardId instead of a text. In both assertion if the the system fails the to return on the dashboard page, the test must fail!

**Comment by kdaud** [comment]
@sherrif10 any more reviews to address?

File: src/test/java/org/openmrs/contrib/qaframework/automation/ConditionsSteps.java (hunk 2 above)
Comment (sherrif10) [reply src/test/java/org/openmrs/contrib/qaframework/automation/ConditionsSteps.java]: Great, so ideally you are saying the best approach would assertPage instead of text on the page, i do believe this approach also does the same thing, However i havent seen any logic behind this kind of approach, since we might populate assertPage in assertion area, this might look like not clean code, i do believe using text() still would bring the implementation more sufficient, Am not objecting the idea, but am trying to look at the idea that is going to be more appropriate

Comment (kdaud) [reply src/test/java/org/openmrs/contrib/qaframework/automation/ConditionsSteps.java]: Since we are dealing with loading pages, it would better to rather affirm the page is loaded  successfully using `assertPage() `than using the `assertText()` on the page. If we use texts, then the tests will fail the run in case;
1. the text in the assertion has been changed by UI designers.
2. the locale is changed from English.

File: src/test/java/org/openmrs/contrib/qaframework/automation/AllergiesSteps.java [hunk 3]
```diff
@@ -42,13 +42,13 @@ public void loadAllergiesPage() {
...
 	public void systemLoadsAllergiesPage() {
-		assertEquals(getElement(patientHeaderId).getText(),getElement(patientHeaderId).getText());
-		assertTrue(textExists("Allergies"));
+		assertPage(allergyPage.waitForPage());
 	}
 
 	@And("a user clicks No Known Allergy button")
 	public void addNoKnownAllergy() {
 		allergyPage.addNoKnownAllergy();
+		allergyPage.waitForPage();
```
Comment (sherrif10) [inline src/test/java/org/openmrs/contrib/qaframework/automation/AllergiesSteps.java:51]: is this line neccesary here

File: src/test/java/org/openmrs/contrib/qaframework/automation/AllergiesSteps.java [hunk 4]
```diff
@@ -59,6 +59,7 @@ public void systemAddsNoKnownAllergies() {
 	@And("a user clicks Remove No Known Allergy icon")
 	public void removeNoKnownAllergy() {
 		allergyPage.removeNoKnownAllergy();
+		allergyPage.waitForPage();
```
Comment (sherrif10) [inline src/test/java/org/openmrs/contrib/qaframework/automation/AllergiesSteps.java:62]: Same here

File: src/test/java/org/openmrs/contrib/qaframework/automation/PageRejectedException.java [hunk 5]
```diff
@@ -27,5 +27,4 @@ public PageRejectedException(String message, Throwable cause,
 			boolean enableSuppression, boolean writableStackTrace) {
 		super(message, cause, enableSuppression, writableStackTrace);
 	}
-
```
Comment (sherrif10) [inline src/test/java/org/openmrs/contrib/qaframework/automation/PageRejectedException.java:30]: Do we also need empty line here

File: src/test/java/org/openmrs/contrib/qaframework/automation/AllergiesSteps.java (hunk 4 above)
Comment (kdaud) [reply src/test/java/org/openmrs/contrib/qaframework/automation/AllergiesSteps.java:62]: adding/removing the allergy some times takes longer and often fails via Bamboo. Invoking waitForPage() gives a room for the operations to be successful before the test performs the next action. Have been observing Bamboo builds failures concerning these operations and we would need all the flaws within these tests fixed so that we the module is stable.

File: src/test/java/org/openmrs/contrib/qaframework/automation/AllergiesSteps.java (hunk 3 above)
Comment (kdaud) [reply src/test/java/org/openmrs/contrib/qaframework/automation/AllergiesSteps.java:51]: refer to my comment [here](https://github.com/openmrs/openmrs-contrib-qaframework/pull/256#discussion_r763906928)

File: src/test/java/org/openmrs/contrib/qaframework/automation/PageRejectedException.java (hunk 5 above)
Comment (kdaud) [reply src/test/java/org/openmrs/contrib/qaframework/automation/PageRejectedException.java:30]: > Do we also need empty line here

we do not need it and its why is removed!

**Review by sherrif10 — APPROVED** [review]
Thanks @kdaud

File: src/test/java/org/openmrs/contrib/qaframework/automation/ConditionsSteps.java (hunk 2 above)
Comment (dkayiwa) [reply src/test/java/org/openmrs/contrib/qaframework/automation/ConditionsSteps.java]: @kdaud what happens when the page loads without the expected test? Will the test still fail as expected?

File: src/test/java/org/openmrs/contrib/qaframework/automation/AllergiesSteps.java (hunk 3 above)
Comment (dkayiwa) [reply src/test/java/org/openmrs/contrib/qaframework/automation/AllergiesSteps.java:51]: @kdaud which action fails when you do not wait? I thought you would wait before the action that fails. Not after the previous one.

**Comment by dkayiwa** [comment]
@kdaud i have done a quick review of this. But am more inclined to @sherrif10 's feelings about these changes. So would need a convincing explanation to think otherwise.

File: src/test/java/org/openmrs/contrib/qaframework/automation/ConditionsSteps.java (hunk 2 above)
Comment (kdaud) [reply src/test/java/org/openmrs/contrib/qaframework/automation/ConditionsSteps.java]: > @kdaud what happens when the page loads without the expected test? Will the test still fail as expected?

Have stepped back over the question and realistically the test will pass without the expected test. This would mean resorting to asserting texts on the page however, am skeptical  over asserting tests via texts, in case the locale is changed from English all the tests will fail. Unless we assume this scenario not to happen any way.

File: src/test/java/org/openmrs/contrib/qaframework/automation/AllergiesSteps.java (hunk 3 above)
Comment (kdaud) [reply src/test/java/org/openmrs/contrib/qaframework/automation/AllergiesSteps.java:51]: adding known allergy seems to take longer to perform the action so as the next operation to be effective.

**Comment by kdaud** [comment]
@dkayiwa @sherrif10 your thoughts are worth! The middle ground is to have these tests hard enough not to break so as to serve the intended purpose. Let me know about you thoughts before we may qualify this commit for a revert action.

**Comment by sherrif10** [comment]
@kdaud. To make things more precise, I would assert text instead of a
waiting of a page. These make even the reader to clearly telll what is the
expectations.

File: src/test/java/org/openmrs/contrib/qaframework/automation/ConditionsSteps.java (hunk 2 above)
Comment (dkayiwa) [reply src/test/java/org/openmrs/contrib/qaframework/automation/ConditionsSteps.java]: Is asserting text the only way we can reliably locate it?

Comment (kdaud) [reply src/test/java/org/openmrs/contrib/qaframework/automation/ConditionsSteps.java]: @sherrif10 is there an alternative for asserting texts for loading system page?

Comment (kdaud) [reply src/test/java/org/openmrs/contrib/qaframework/automation/ConditionsSteps.java]: > Is asserting text the only way we can reliably locate it?

@dkayiwa have not yet known another way but hope will get a better way, for now we can resort to asserting texts.

**Comment by kdaud** [comment]
will revert this commit before Friday ends in case no one objects.
cc: @sherrif10  @dkayiwa

**Comment by dkayiwa** [comment]
No objection from me.

**Comment by sherrif10** [comment]
Me too

**Comment by kdaud** [comment]
Have just noticed that the changes in this commit do not require Revert action since they were made in duplicate files which are being addressed in `pull/265`, so they won't appear in the repo once the pull is merged !!

**Comment by sherrif10** [comment]
> duplicate files which are being addressed in `pull/265`,

Is this already merged or still in progress.

**Comment by dkayiwa** [comment]
> they were made in duplicate files

Here comes another problem of duplicating files. ☺️

**Comment by kdaud** [comment]
> Is this already merged or still in progress

still in progress

**Comment by kdaud** [comment]
> Here comes another problem of duplicating files

Its another C-19 of type. We have other duplicates in other repos that need to address ASAP cc: @sherrif10

**Comment by Irenyak1** [comment]
@kdaud , @sherrif10 ,  @dkayiwa ,  from this [comment](https://github.com/openmrs/openmrs-contrib-qaframework/pull/256#issuecomment-988907267), do we go with asserting text as opposed to pages? Is this our agreed stand?
